### PR TITLE
fix(pipeline): The repair cron task cannot be turned on or off

### PR DESCRIPTION
**Pull request by kakj-go**
#### What this PR does / why we need it:
The repair cron task cannot be turned on or off

#### Which issue(s) this PR fixes:

- [Erda Cloud Issue Link](https://erda.cloud/erda/dop/projects/387/issues/all?id=297877&issueFilter__urlQuery=eyJzdGF0ZXMiOls0NDEyLDQ1MzgsNDQxMyw0NDE0LDQ0MTUsNDQxNl0sImFzc2lnbmVlSURzIjpbIjEwMDA1NjAiXX0%3D&issueTable__urlQuery=eyJwYWdlTm8iOjEsICJwYWdlU2l6ZSI6MTB9&iterationID=1146&type=BUG)

#### Specified Reviewers:

/assign @your-reviewer

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |      The repair cron task cannot be started normally        |
| 🇨🇳 中文    |       修复定时任务无法正常启动       |

### CODE REVIEW

**Comment by kakj-go** [comment]
test image: 

Cron normal scheduling

image

Cron normal scheduling done
image

Implement compensation strategy
image

File: modules/pipeline/providers/cron/daemon/cron_task.go [hunk 1]
```diff
@@ -158,72 +158,66 @@ func (d *provider) DoCrondAbout(ctx context.Context) {
 // Listen the key in the channel
 // decide to delete or add a scheduled task according to the prefix of the key
 func (s *provider) listenCrond(ctx context.Context) {
+	rch := s.EtcdClient.Watch(ctx, etcdCronWatchPrefix, clientv3.WithPrefix())
```
Comment (sfwn) [inline modules/pipeline/providers/cron/daemon/cron_task.go:161]: use lw.ListenPrefix

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/providers/cron/daemon/cron_task.go (hunk 1 above)
Comment (kakj-go) [reply modules/pipeline/providers/cron/daemon/cron_task.go]: done

File: modules/pipeline/providers/cron/daemon/cron_task.go [hunk 2]
```diff
@@ -158,75 +158,59 @@ func (d *provider) DoCrondAbout(ctx context.Context) {
...
 // decide to delete or add a scheduled task according to the prefix of the key
 func (s *provider) listenCrond(ctx context.Context) {
-	for {
-		select {
-		case <-ctx.Done():
+	s.LeaderWorker.ListenPrefix(ctx, etcdCronWatchPrefix, func(ctx context.Context, event *clientv3.Event) {
+		t := event.Type
+		key := string(event.Kv.Key)
+
+		logrus.Infof("crond: watched update etcd key: %s, changeType: %d", key, t)
```
Comment (sfwn) [inline modules/pipeline/providers/cron/daemon/cron_task.go:165]: use s.Log

File: modules/pipeline/providers/cron/daemon/cron_task.go [hunk 3]
```diff
@@ -158,75 +158,59 @@ func (d *provider) DoCrondAbout(ctx context.Context) {
...
+
+		// if change type is delete, don't need do anything
+		if t == mvccpb.DELETE {
```
Comment (sfwn) [inline modules/pipeline/providers/cron/daemon/cron_task.go:168]: this is put handler, action never be DELETE

File: modules/pipeline/providers/cron/daemon/cron_task.go [hunk 4]
```diff
@@ -158,75 +158,59 @@ func (d *provider) DoCrondAbout(ctx context.Context) {
...
+			return
+		}
+
+		if _, err := s.EtcdClient.Delete(ctx, key); err != nil {
+			logrus.Errorf("crond: failed to delete key: %s", key)
```
Comment (sfwn) [inline modules/pipeline/providers/cron/daemon/cron_task.go:173]: s.Log

**Review by sfwn — CHANGES_REQUESTED** [review]

**Review by sfwn — APPROVED** [review]

**Pull request merged**